### PR TITLE
[µTVM] Fix problems with the debug flow

### PR DESCRIPTION
**Pull request by areusch**
This commit makes an internal API change in the µTVM transport library and which allows for some fixes to the GdbDebugger class, improving if not unbreaking the debug flow. When timeouts were introduced to the µTVM transport, it was assumed that a timeout was always wanted, since a lack of timeout could break automated tests. However, in debugging, this is is not the case--in fact, no timeouts are wanted, save on physical hardware where a session start retry timeout can be useful.

Overall effects of this change:
 - When tvm.micro.transport.Transport.read() or tvm.micro.transport.Transport.write() are called, timeout_sec can now be None. When None, implementations should block forever. Previously, 0 could have meant "block forever" (almost always) or "perform non-blocking I/O" (in edge cases on session start). Now, there is an unambiguous meaning that matches pyserial.
 - When using GdbDebugger, never terminate the debugger. Always block on the debugger until the user exits. This allows the user to finish debugging problems without stepping on them in the event that a host-side exception occurs.
 - Immediately restore Ctrl+C handler when the debugger quits. If the debugger cleanup flow hangs, allows the user to typically find the problem by Ctrl+C. Before, explicit kill was required.

### CODE REVIEW

**Comment by areusch** [comment]
@tqchen @tom-gall @u99127 @manupa-arm 

Note: this mostly impacts the x86 flow used with test_crt, but submitting this to make things not broken and more maintainable. One thing that may impact the zephyr flow is the "debugger not kill'd on error."

File: python/tvm/micro/debugger.py [hunk 1]
```diff
@@ -82,23 +73,56 @@ def _stop_all(cls):
...
     def _wait_for_child(self):
         self.popen.wait()
-        with self._child_alive_lock:
-            self._is_child_alive = True
+        with self._is_running_lock:
+            self._internal_stop()
+
+    @classmethod
+    def _sigusr1_handler(cls, signum, stack_frame):  # pylint: disable=unused-argument
+        if cls._STARTED_INSTANCE is not None:
```
Comment (liangfu) [inline python/tvm/micro/debugger.py:113]: Just curious, why not raise an error when STARTED_INSTANCE is None?

File: python/tvm/micro/debugger.py [hunk 2]
```diff
@@ -82,23 +73,56 @@ def _stop_all(cls):
...
+            signal.signal(signal.SIGINT, cls._STARTED_INSTANCE.old_sigint_handler)
+            signal.signal(signal.SIGUSR1, cls._STARTED_INSTANCE.old_sigusr1_handler)
+            cls._STARTED_INSTANCE._signals_reset_event.set()
+            return
+
+        raise Exception()
```
Comment (liangfu) [inline python/tvm/micro/debugger.py:119]: Please leave some meaningful message if possible.

File: python/tvm/micro/debugger.py [hunk 3]
```diff
@@ -109,46 +133,32 @@ def _sigint_handler(cls, signum, stack_frame):  # pylint: disable=unused-argumen
         raise Exception()
```
Comment (liangfu) [inline python/tvm/micro/debugger.py:133]: mind leave a message here as well?

**Review by liangfu — APPROVED** [review]
Thanks for your contribution. The debug flow looks good. I just have the nitpicks.

File: python/tvm/micro/debugger.py (hunk 1 above)
Comment (areusch) [reply python/tvm/micro/debugger.py:116]: good call, i've inverted the logic and made the check stricter.

File: python/tvm/micro/debugger.py (hunk 2 above)
Comment (areusch) [reply python/tvm/micro/debugger.py:122]: done

File: python/tvm/micro/debugger.py (hunk 3 above)
Comment (areusch) [reply python/tvm/micro/debugger.py:136]: done

**Comment by areusch** [comment]
thanks @liangfu I found a few lines of bad merge (this was pulled out of the linked params PR #6917) but have re-tested locally with changes from your comments and it looks good now

File: python/tvm/micro/debugger.py (hunk 1 above)
Comment (liangfu) [reply python/tvm/micro/debugger.py:116]: the logic seems not inverted yet. missed something in the submit?

Comment (areusch) [reply python/tvm/micro/debugger.py]: oooops good catch. actually fixed that now.

**Pull request merged**